profiler/internal/cmemprof/callers.go

```diff
@@ -0,0 +1,36 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022 Datadog, Inc.
+
+package cmemprof
+
+import (
+	"runtime"
+)
+
+/*
+#include <stdint.h> // uintptr_t
+*/
+import "C"
+
+//export recordAllocationSample
+func recordAllocationSample(size uint) {
+	p := activeProfile.Load().(*Profile)
+	if p == nil {
+		return
+	}
+
+	// There are several calls in the call stack we should skip. Ultimiately
+	// we'd like to get only the actual allocation call, but we can settle
+	// for profile_allocation. This call should be skipped, together with
+	// any of the C->Go transition functions like cgocallback and the
+	// wrapper for this function that's called through the exported version.
+	//
+	// TODO: how much to skip? Is the value consistent across Go releases?
+	// The value below just comes from experimentation with Go 1.17.5
+	const skip = 6
```

Comment:
It's implicitly tested by `TestCAllocationProfiler` because the expected samples won't show up if `skip` is too high. But I agree that it would be good to have some kind of way to check that we're not skipping too little, either. The frames that are getting skipped are mostly the C -> Go cross-calling functions. That exact sequence of functions [has changed in the past](https://go-review.googlesource.com/c/go/+/258938) and could change again.